opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs

```diff
@@ -0,0 +1,410 @@
+                    Value::IntValue(_) => BondDataType::BT_INT32 as u8,
+                    Value::DoubleValue(_) => BondDataType::BT_FLOAT as u8, // TODO - using float for now
+                    Value::BoolValue(_) => BondDataType::BT_INT32 as u8, // representing bool as int
+                    _ => continue,
+                };
+                fields.push((Cow::Owned(attr.key.clone()), type_id));
+            }
+        }
+        fields.sort_by(|a, b| a.0.cmp(&b.0)); // Sort fields by name consistent schema ID generation
+        fields
+            .into_iter()
+            .enumerate()
+            .map(|(i, (name, type_id))| FieldDef {
+                name,
+                type_id,
+                field_id: (i + 1) as u16,
+            })
+            .collect()
+    }
+
+    /// Calculate schema ID from field specifications
+    fn calculate_schema_id(fields: &[FieldDef]) -> u64 {
+        use std::collections::hash_map::DefaultHasher;
+        use std::hash::{Hash, Hasher};
+
+        let mut hasher = DefaultHasher::new();
+
+        for field in fields {
+            field.name.hash(&mut hasher);
+            field.type_id.hash(&mut hasher);
+        }
+
+        hasher.finish()
+    }
+
+    /// Get or create schema with field ordering information
+    fn get_or_create_schema(
+        &self,
+        schema_id: u64,
+        field_info: Vec<FieldDef>,
+    ) -> (CentralSchemaEntry, Vec<FieldDef>) {
+        // Check cache first
+        if let Some((schema, schema_md5, field_info)) =
+            self.schema_cache.read().unwrap().get(&schema_id)
+        {
+            return (
+                CentralSchemaEntry {
+                    id: schema_id,
+                    md5: *schema_md5,
+                    schema: schema.clone(),
+                },
+                field_info.clone(),
+            );
+        }
+
+        let schema =
+            BondEncodedSchema::from_fields("OtlpLogRecord", "telemetry", field_info.clone()); //TODO - use actual struct name and namespace
+
+        let schema_bytes = schema.as_bytes();
+        let schema_md5 = md5::compute(schema_bytes).0;
+        // Cache the schema and field info
+        {
+            let mut cache = self.schema_cache.write().unwrap();
+            // TODO: Refactor to eliminate field_info duplication in cache
+            // The field information (name, type_id, order) is already stored in BondEncodedSchema's
+            // DynamicSchema.fields vector. We should:
+            // 1. Ensure DynamicSchema maintains fields in sorted order
+            // 2. Add a method to BondEncodedSchema to iterate fields for row encoding
+            // 3. Remove field_info from cache tuple to reduce memory usage and cloning overhead
+            // This would require updating write_row_data() to work with DynamicSchema fields directly
+            cache.insert(schema_id, (schema.clone(), schema_md5, field_info.clone()));
```

Comment:
Use double-checked locking to avoid updating the entry if one already got added.